pkgs/applications/editors/emacs/elisp-packages/manual-packages/lua/package.nix

```
@@ -0,0 +1,42 @@
+{
+  lib,
+  stdenv,
+  lua,
+  melpaBuild,
+  pkg-config,
+  fetchFromGitHub,
+  unstableGitUpdater,
+}:
+
+melpaBuild {
+  pname = "lua";
+  version = "0-unstable-2025-01-27";
+
+  src = fetchFromGitHub {
+    owner = "syohex";
+    repo = "emacs-lua";
+    rev = "501189b5fc069fcead8843b2b0ad510c08de1397";
+    hash = "sha256-psCrto12p03R9XxPtDYTMB5vcRVWj+Blq7D30nLsSbU=";
+  };
+
+  preBuild = ''
+    make LUA_VERSION=${lua.luaversion}
+  '';
+
+  ignoreCompilationError = false;
+
+  nativeBuildInputs = [ pkg-config ];
+
+  buildInputs = [ lua ];
+
+  files = ''(:defaults "lua-core${stdenv.hostPlatform.extensions.sharedLibrary}")'';
```

```suggestion
  files = ''(:defaults "lua-core.so")'';
```

This should be enough because `.so` is also [supported] on darwin.

If you still want to use `.dylib`, you have to patch upstream `Makefile` because it [hardcodes] `.so`.

[supported]: https://www.gnu.org/software/emacs/manual/html_node/elisp/Dynamic-Modules.html
[hardcodes]: https://github.com/syohex/emacs-lua/blob/501189b5fc069fcead8843b2b0ad510c08de1397/Makefile#L15-L16